UnifiedPDF: Paint wash and holes for find matches

hortont424 opened the pull request:
#### 9178642d517556d27941c905e8a8255af40fcc71
<pre>
UnifiedPDF: Paint wash and holes for find matches
<a href="https://bugs.webkit.org/show_bug.cgi?id=269493">https://bugs.webkit.org/show_bug.cgi?id=269493</a>
<a href="https://rdar.apple.com/123028742">rdar://123028742</a>

Reviewed by Simon Fraser.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::rectsForTextMatchesInRect):
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::updateFindUIAfterPageScroll):
(WebKit::FindController::rectsForTextMatchesInRect):
(WebKit::FindController::didInvalidateFindRects):
(WebKit::FindController::didInvalidateDocumentMarkerRects): Deleted.
* Source/WebKit/WebProcess/WebPage/FindController.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::didInvalidateDocumentMarkerRects):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::rectsForTextMatchesInRect):
(WebKit::PluginView::drawsFindOverlay const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
Have FindController defer to the plugin as to whether it wants the find overlay,
and to retrieve the rectangles for the holes.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::pdfPageRectToDocumentRect const):
Actually flip this rectangle, instead of just moving it.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):
Invalidate find rects when we scroll. For now they&apos;ll scroll asynchronously
in the main find overlay, but we should move them into an overlay layer that\
scrolls with the PDF instead.

(WebKit::compareOptionsForFindOptions):
Factor this conversion out of findString.

(WebKit::UnifiedPDFPlugin::findString):
When the find string changes, rebuild the cached set of match rects.

(WebKit::UnifiedPDFPlugin::collectFindMatchRects):
Maintain a set of rects for all of the matches, in document coordinates.

(WebKit::UnifiedPDFPlugin::rectsForTextMatchesInRect):
Convert and return the rects in plugin coordinates. This only works for
the main frame plugin, but FindController is only hooked up for the main
frame plugin for now.

(WebKit::UnifiedPDFPlugin::pluginBoundsForAnnotation const):
Un-work-around the missing flip in pdfPageRectToDocumentRect.

Canonical link: <a href="https://commits.webkit.org/274781@main">https://commits.webkit.org/274781@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35b501ffd51a5f602438dede92eee20116e93565

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40008 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42384 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42553 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16349 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40582 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/16020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43831 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39634 "Build is in progress. Recent messages:") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/16442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8978 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2410,6 +2424,28 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
+    m_frame->page()->findController().didInvalidateFindRects();
+}
+
+Vector<FloatRect> UnifiedPDFPlugin::rectsForTextMatchesInRect(IntRect clipRect)
```
Can be const

smfr replied:
We pass rects by reference, right?

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -961,6 +962,9 @@ static String mutationObserverNotificationString()
     determineCurrentlySnappedPage();
 
+    // FIXME: Make the overlay scroll with the tiles instead of repainting constantly.
+    m_frame->page()->findController().didInvalidateFindRects();
```
`protectedPage()`

hortont424 replied:
Good point! This does not match all the others.

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2362,29 +2381,25 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
-        if (!isCaseSensitive)
-            options |= NSCaseInsensitiveSearch;
-        PDFSelection *foundSelection = [m_pdfDocument findString:target fromSelection:m_currentSelection.get() withOptions:options];
+        PDFSelection *foundSelection = [m_pdfDocument findString:target fromSelection:m_currentSelection.get() withOptions:compareOptions];
```
Can this use `RetainPtr` ?

hortont424 replied:
Happily

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2410,6 +2425,28 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
+        }
+    }
+
+    m_frame->page()->findController().didInvalidateFindRects();
```
`protectedPage()` as it owns the FindController and you want to make sure it stays alive during the whole `didInvalidateFindRects()` execution.

hortont424 replied:
Sure!

cdumez commented on Source/WebKit/WebProcess/Plugins/PluginView.cpp:
```diff
@@ -595,6 +595,22 @@ bool PluginView::findString(const String& target, WebCore::FindOptions options,
+    if (!m_isInitialized)
+        return { };
+
+    return m_plugin->rectsForTextMatchesInRect(clipRect);
```
Please protect `m_plugin`. Ideally, you'd add a `protectedPlugin()` to this class which returns a `Ref<PDFPluginBase>` and call this here.

hortont424 replied:
Iiiiiiiiii think maybe we should do this whole mechanism in one patch? I can look into that.

cdumez commented on Source/WebKit/WebProcess/Plugins/PluginView.cpp:
```diff
@@ -595,6 +595,22 @@ bool PluginView::findString(const String& target, WebCore::FindOptions options,
+    if (!m_isInitialized)
+        return false;
+
+    return m_plugin->drawsFindOverlay();
```
`protectedPlugin()`

cdumez commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -521,6 +521,11 @@ void FindController::redraw()
 Vector<FloatRect> FindController::rectsForTextMatchesInRect(IntRect clipRect)
 {
+#if ENABLE(PDF_PLUGIN)
+    if (auto* pluginView = mainFramePlugIn())
```
`auto*` -> `RefPtr`

cdumez reviewed (commented):
Please (please please) get in the habit of using smart pointers more.

hortont424 committed:
UnifiedPDF: Paint wash and holes for find matches
https://bugs.webkit.org/show_bug.cgi?id=269493
rdar://123028742

Reviewed by Simon Fraser.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::rectsForTextMatchesInRect):
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::updateFindUIAfterPageScroll):
(WebKit::FindController::rectsForTextMatchesInRect):
(WebKit::FindController::didInvalidateFindRects):
(WebKit::FindController::didInvalidateDocumentMarkerRects): Deleted.
* Source/WebKit/WebProcess/WebPage/FindController.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::didInvalidateDocumentMarkerRects):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::rectsForTextMatchesInRect):
(WebKit::PluginView::drawsFindOverlay const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
Have FindController defer to the plugin as to whether it wants the find overlay,
and to retrieve the rectangles for the holes.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::pdfPageRectToDocumentRect const):
Actually flip this rectangle, instead of just moving it.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):
Invalidate find rects when we scroll. For now they'll scroll asynchronously
in the main find overlay, but we should move them into an overlay layer that\
scrolls with the PDF instead.

(WebKit::compareOptionsForFindOptions):
Factor this conversion out of findString.

(WebKit::UnifiedPDFPlugin::findString):
When the find string changes, rebuild the cached set of match rects.

(WebKit::UnifiedPDFPlugin::collectFindMatchRects):
Maintain a set of rects for all of the matches, in document coordinates.

(WebKit::UnifiedPDFPlugin::rectsForTextMatchesInRect):
Convert and return the rects in plugin coordinates. This only works for
the main frame plugin, but FindController is only hooked up for the main
frame plugin for now.

(WebKit::UnifiedPDFPlugin::pluginBoundsForAnnotation const):
Un-work-around the missing flip in pdfPageRectToDocumentRect.

Canonical link: https://commits.webkit.org/274781@main
